thetis/momentum_eq.py

```diff
@@ -297,31 +300,80 @@ def rhs(self, solution, eta, w=None, viscosity_v=None,
         if viscosity_h is not None:
-            f_visc = viscosity_h * (Dx(solution[0], 0) * Dx(self.test[0], 0) +
-                                    Dx(solution[1], 0) * Dx(self.test[1], 0) +
-                                    Dx(solution[0], 1) * Dx(self.test[0], 1) +
-                                    Dx(solution[1], 1) * Dx(self.test[1], 1))
+            def grad_h(a):
+                return as_matrix([[Dx(a[0], 0), Dx(a[0], 1), 0],
+                                  [Dx(a[1], 0), Dx(a[1], 1), 0],
+                                  [0, 0, 0]])
+
+            def outer_jump(v, n):
+                return outer(v('+'), n('+')) + outer(v('-'), n('-'))
+            visc_tensor = as_matrix([[viscosity_h, 0, 0],
+                                     [0, viscosity_h, 0],
+                                     [0, 0, 0]])
+
+            grad_uv = grad_h(solution)
+            grad_test = grad_h(self.test)
+            stress = visc_tensor*grad_uv
+            f += inner(grad_test, stress)*dx
+
             if self.horizontal_dg:
-                # interface term
-                mu_grad_sol = viscosity_h*nabla_grad(solution)
-                f += -(avg(mu_grad_sol[0, 0])*jump(self.test[0], self.normal[0]) +
-                       avg(mu_grad_sol[0, 1])*jump(self.test[1], self.normal[0]) +
-                       avg(mu_grad_sol[1, 0])*jump(self.test[0], self.normal[1]) +
-                       avg(mu_grad_sol[1, 1])*jump(self.test[1], self.normal[1]))*(dS_v+dS_h)
-                # TODO symmetric interior penalty term
-            f += f_visc * dx
+                assert self.h_elem_size is not None, 'h_elem_size must be defined'
+                # Interior Penalty method by
+                # Epshteyn (2007) doi:10.1016/j.cam.2006.08.029
+                # sigma = 3*k_max**2/k_min*p*(p+1)*cot(Theta)
+                # k_max/k_min  - max/min diffusivity
+                # p            - polynomial degree
+                # Theta        - min angle of triangles
+                # assuming k_max/k_min=2, Theta=pi/3
+                # sigma = 6.93 = 3.5*p*(p+1)
+                degree_h, degree_v = self.space.ufl_element().degree()
+                # TODO compute elemsize as CellVolume/FacetArea
+                # h = n.D.n where D = diag(h_h, h_h, h_v)
+                elemsize = (self.h_elem_size*(self.normal[0]**2 +
+                                              self.normal[1]**2) +
+                            self.v_elem_size*self.normal[2]**2)
+                sigma = 3.5*degree_h*(degree_h + 1)/elemsize
+                if degree_h == 0:
+                    raise NotImplementedError('horizontal visc not implemented for p0')
+                alpha = avg(sigma)
+                ds_interior = (dS_h + dS_v)
+                # oriented diffusive flux n.K.n = nu*(n_xy.n_xy)
+                # Pestiaux (2014)  DOI: 10.1002/fld.3900
+                ip_flux_jump = avg(dot(dot(self.normal, visc_tensor), self.normal))*jump(solution)
```

Comment:
Thanks a lot for checking this. Your formulations look legit, I'll go through them more carefully and the implementation.